### PR TITLE
Do not set header X-Goog-User-Project header for the resource google_client_openid_userinfo

### PR DESCRIPTION
**Pull request by zli82016**
Fixes https://github.com/hashicorp/terraform-provider-google/issues/10260

https://b.corp.google.com/issues/202287364

When setting the header `X-Goog-User-Project` for [the call to get user info](https://github.com/hashicorp/terraform-provider-google/blob/release-3.81.0/google/data_source_google_client_openid_userinfo.go#L28), it returns an error. Maybe the API openidconnect.googleapis.com does not handle the header  `X-Goog-User-Project` properly. The error does not occur with other APIs.  So don't set the header `X-Goog-User-Project` for the resource `google_client_openid_userinfo`. 

If this PR is for Terraform, I acknowledge that I have:

- [X] Searched through the [issue tracker](https://github.com/hashicorp/terraform-provider-google/issues) for an open issue that this either resolves or contributes to, commented on it to claim it, and written "fixes {url}" or "part of {url}" in this PR description. If there were no relevant open issues, I opened one and commented that I would like to work on it (not necessary for very small changes).
- [X] [Generated Terraform](https://github.com/GoogleCloudPlatform/magic-modules#generating-the-terraform-providers), and ran `make test` and `make lint` to ensure it passes unit and linter tests.
- [X] Ensured that all new fields I added that can be set by a user appear in at least one [example](https://github.com/GoogleCloudPlatform/magic-modules/tree/main/mmv1/templates/terraform/examples) (for generated resources) or [third_party test](https://github.com/GoogleCloudPlatform/magic-modules/tree/main/mmv1/third_party/terraform/tests) (for handwritten resources or update tests).
- [X] [Ran](https://github.com/hashicorp/terraform-provider-google/blob/main/.github/CONTRIBUTING.md#tests) relevant acceptance tests (If the acceptance tests do not yet pass or you are unable to run them, please let your reviewer know).
- [X] Read the [Release Notes Guide](https://github.com/GoogleCloudPlatform/magic-modules/blob/main/.ci/RELEASE_NOTES_GUIDE.md) before writing my release note below.

**Release Note Template for Downstream PRs (will be copied)**

```release-note:bug
cloudplatform: fixed the error with header `X-Goog-User-Project` on `google_client_openid_userinfo`
```

### CODE REVIEW

File: mmv1/third_party/terraform/utils/transport.go.erb [hunk 1]
```diff
@@ -50,7 +50,9 @@ func sendRequestWithTimeout(config *Config, method, project, rawurl, userAgent s
 	reqHeaders.Set("User-Agent", userAgent)
 	reqHeaders.Set("Content-Type", "application/json")
 
-	if config.UserProjectOverride && project != "" {
+	if !config.UserProjectOverride {
+		reqHeaders.Set("X-Goog-User-Project", "")
```
Comment (melinath) [inline mmv1/third_party/terraform/utils/transport.go.erb:54]: Why is this necessary?

**Review by melinath — COMMENTED** [review]
It looks like there are a few different places that GetCurrentUserEmail is called - do they all have this problem? If so - could this workaround be shifted inside GetCurrentUserEmail?

**Comment by zli82016** [comment]
> It looks like there are a few different places that GetCurrentUserEmail is called - do they all have this problem? If so - could this workaround be shifted inside GetCurrentUserEmail?

Actually, only data source has the problem. Other places where GetCurrentUserEmail are used do not have the problem because the c.UserProjectOverride && c.BillingProject is not set in the config.

Do we still need to move the workaround inside GetCurrentUserEmail incase GetCurrentUserEmail will be used for more places in the future?

File: mmv1/third_party/terraform/utils/transport.go.erb (hunk 1 above)
Comment (melinath) [reply mmv1/third_party/terraform/utils/transport.go.erb:54]: What is the value of explicitly setting the "X-Goog-User-Project" header to empty? It seems like the current behavior should be to not set it at all, which I would expect to be enough - do we also need to make sure that it gets unset if it was previously set somewhere else?

**Review by melinath — CHANGES_REQUESTED** [review]
> Do we still need to move the workaround inside GetCurrentUserEmail incase GetCurrentUserEmail will be used for more places in the future?

It sounds like GetCurrentUserEmail should never trigger billing - so I think it would be better to move the workaround inside the function. Even if none of the usage currently has this issue, it would be easy to accidentally make a change or use this somewhere else and run into this same issue. Let's make sure it's always consistent instead.

File: mmv1/third_party/terraform/utils/transport.go.erb (hunk 1 above)
Comment (zli82016) [reply mmv1/third_party/terraform/utils/transport.go.erb:54]: Current behavior is that when environment variables `USER_PROJECT_OVERRIDE` is set to true and `GOOGLE_BILLING_PROJECT` is set to some project, "X-Goog-User-Project" header is set to `GOOGLE_BILLING_PROJECT`. https://github.com/hashicorp/terraform-provider-google/blob/26ff814b39d07b56b657eeb92e0abf423077be58/google/config.go#L396-L398

	if c.UserProjectOverride && c.BillingProject != "" {
		headerTransport.Set("X-Goog-User-Project", c.BillingProject)
	}

The issue is that when "X-Goog-User-Project" header is set to `GOOGLE_BILLING_PROJECT`, an error occurs for the resource `google_client_openid_userinfo` specifically and not for other resources. To avoid the error,  I set  UserProjectOverride to false before the API request to get the user info 

	origUserProjectOverride := config.UserProjectOverride
	config.UserProjectOverride = false

and then the header "X-Goog-User-Project" is set to empty with false UserProjectOverride.
	
     if !config.UserProjectOverride {
		reqHeaders.Set("X-Goog-User-Project", "")
    }


To not affect other resources,   reset UserProjectOverride to original value after the API request for the resource `google_client_openid_userinfo`. 

       config.UserProjectOverride = origUserProjectOverride

**Comment by zli82016** [comment]
> > Do we still need to move the workaround inside GetCurrentUserEmail incase GetCurrentUserEmail will be used for more places in the future?
> 
> It sounds like GetCurrentUserEmail should never trigger billing - so I think it would be better to move the workaround inside the function. Even if none of the usage currently has this issue, it would be easy to accidentally make a change or use this somewhere else and run into this same issue. Let's make sure it's always consistent instead.

That makes sense. Moved the workaround inside GetCurrentUserEmail.

File: mmv1/third_party/terraform/utils/transport.go.erb (hunk 1 above)
Comment (zli82016) [reply mmv1/third_party/terraform/utils/transport.go.erb:54]: > If config.UserProjectOverride is set to false, the block calling reqHeaders.Set("X-Goog-User-Project", project) will already be skipped, which means that the header won't be sent at all. That would be the same behavior that calls to GetCurrentUserEmail with no project set currently have in the provider, because they already aren't setting the header at all.

Actually when the environment variable `USER_PROJECT_OVERRIDE` is set to true and `GOOGLE_BILLING_PROJECT` is set to some project, first the header `X-Goog-User-Project` is set in provider config. 
https://github.com/GoogleCloudPlatform/magic-modules/blob/main/mmv1/third_party/terraform/utils/config.go.erb#L307-L311


And then for some resources, the header `X-Goog-User-Project` is set again in resource level (this block of code). 

      if config.UserProjectOverride && project != "" {
		  // Pass the project into this fn instead of parsing it from the URL because
		  // both project names and URLs can have colons in them.
		  reqHeaders.Set("X-Goog-User-Project", project)
      }

If config.UserProjectOverride is set to false, the block calling reqHeaders.Set("X-Goog-User-Project", project) will already be skipped. That is right. But  the header `X-Goog-User-Project` is already set in the provider level.
To  override the header `X-Goog-User-Project` to empty in resource level, I added the code block

	if !config.UserProjectOverride {
		reqHeaders.Set("X-Goog-User-Project", "")
       }

Comment (melinath) [reply mmv1/third_party/terraform/utils/transport.go.erb:54]: Thank you for the detailed explanation, and for writing it out twice - I see that you explained that above and I misunderstood it. Sorry about that, and thanks again!
